scripts/view-test-artifacts.ts

```diff
@@ -0,0 +1,187 @@
+
+type BuildStatus =
+  | 'FAILED'
+  | 'FAULT'
+  | 'IN_PROGRESS'
+  | 'STOPPED'
+  | 'SUCCEEDED'
+  | 'TIMED_OUT';
+
+const buildStatusToIcon: Record<BuildStatus, string> = {
+  FAILED: '❌',
+  FAULT: '🚫',
+  IN_PROGRESS: '🏃',
+  STOPPED: '➖',
+  SUCCEEDED: '✅',
+  TIMED_OUT: '⏰',
+};
+
+/**
+ * Type modelling info we need to retrieve and display test artifacts in the UI.
+ */
+type TestArtifact = {
+  jobName: string;
+  buildStatus: BuildStatus;
+  artifactLocation?: string;
+};
+
+/**
+ * Utility method to filter out incomplete test artifact definitions.
+ * @param artifact the potentially incomplete artifact to validate
+ * @returns whether the artifact is complete or not
+ */
+const testArtifactIsComplete = (artifact: Partial<TestArtifact>): artifact is TestArtifact => artifact.jobName !== undefined;
+
+const generateIndexFile = (directory: string, artifacts: TestArtifact[]): void => {
+  const createArtifactRow = (artifact: TestArtifact): string => `<tr>
+    <td>${artifact.jobName}</td>
+    <td>${buildStatusToIcon[artifact.buildStatus]}[${artifact.buildStatus}]</td>
+    <td>${artifact.artifactLocation ? `<a href="/${artifact.jobName}">Assets</a>` : ''}</td>
+  </tr>`;
+    
+  const fileContents = `<html>
+<head>
+  <title>Test Artifacts</title>
+  <style>
+    table, th, td {
+      border: 1px solid black;
+    }
+  </style>
+</head>
+<body>
+    <h1>Test Artifact Directory</h1>
+    <table>
+      <tr>
+        <th>Job Name</th>
+        <th>Build Status</th>
+        <th>Assets</th>
+      </tr>
+      ${artifacts.map(createArtifactRow).join('\n')}
+    </table>
+</body>
+</html>`;
+  fs.writeFileSync(path.join(directory, 'index.html'), fileContents);
+};
+
+/**
+ * Given a codebuild batch id, download all the builds for that batch, and return the batch name, build status, and artifact location.
+ * @param batchId the batch to look up.
+ */
+const retrieveArtifactsForBatch = async (batchId: string): Promise<TestArtifact[]> => {
+  const codeBuild = new CodeBuild({ region: 'us-east-1' });
+  const { buildBatches } = await codeBuild.batchGetBuildBatches({ ids: [batchId] }).promise();
+  return (buildBatches || [])
+    .flatMap((batch) =>
+      (batch.buildGroups || []).map((buildGroup) => ({
+        jobName: buildGroup.identifier,
+        buildStatus: buildGroup.currentBuildSummary?.buildStatus,
+        artifactLocation: buildGroup.currentBuildSummary?.primaryArtifact?.location,
+      }) as Partial<TestArtifact>),
+    )
+    .filter(testArtifactIsComplete);
+};
+
+const downloadSingleTestArtifact = async (tempDir: string, artifact: Required<TestArtifact>): Promise<unknown[]> => {
+  const artifactDownloadPath = path.join(tempDir, artifact.jobName);
+  fs.mkdirSync(artifactDownloadPath);
+  const [Bucket, ...rest] = artifact.artifactLocation.split(':::')[1].split('/');
+  const Prefix = rest.join('/');
+  const listObjectsResponse = await s3.listObjects({ Bucket, Prefix }).promise();
```

Comment:
As per the AppSec guidance, we need to enable versioning on the buckets which would need some additional filtering here to only get the latest versions of objects.

Reply:
Sounds good, I'll wait til that change is out and we've had some runs to verify behavior before we move forward w/ that.

Reply:
I've tested this script with versioned buckets and it works. Good to go.